### ISSUE TITLE
Data views: quick-editing

### ISSUE THREAD

**Issue by jameskoster** [WordPress/gutenberg#55101]
### Table and Grid layout
In these layouts, we can use the same UI as the full-screen editor to toggle the visibility of the Inspector panel:

https://github.com/user-attachments/assets/66cfe71c-7906-4614-8e3e-5bd3defbd8eb

### List layout
In List layout it feels important that the preview remains visible. Instead of exposing a dedicated panel, the Inspector contents are loaded into the content frame via drilldown pattern:

https://github.com/user-attachments/assets/81516f30-1160-4683-b9ad-db9a744367f8

---

This feature would be equivalent to the quick edit experience in wp-admin:

<img width="1258" alt="Screenshot 2024-04-24 at 15 50 25" src="https://github.com/WordPress/gutenberg/assets/846565/49c863a4-2e07-4f88-af34-c28298643496">

<details><summary>Original issue</summary>Inline editing has considerable overlap with quick-actions (https://github.com/WordPress/gutenberg/issues/55096). For example most quick-actions for pages (https://github.com/WordPress/gutenberg/issues/55628) could also be accessed by clicking on values in the table directly. So for the initial implementation we might reuse where possible. Hopefully this mockup captures that:

<img src="https://user-images.githubusercontent.com/846565/279426851-3d473c16-8dbd-40a8-917f-56a8ee768fa5.png">

Incidentally the "Author" menu in the above concept should be virtually the same as the one that has already been implemented for filtering, so hopefully some legwork is already done 🤞 :

<img width="520" alt="Screenshot 2023-11-17 at 14 00 21" src="https://github.com/WordPress/gutenberg/assets/846565/d5b87906-772f-4c76-a8d7-4cfc847e2b52">

One exception we might make straight away is boolean values, e.g. whether comments are enabled or not. In this case the cell can probably contain a Toggle.</details>

**Comment by jameskoster** [comment]
Noting that the work in https://github.com/WordPress/gutenberg/issues/59689 can unlock using the Inspector as a bulk/quick edit interface:

![Bulk edit](https://github.com/WordPress/gutenberg/assets/846565/58e754e2-8ac1-4ca2-b913-d34b93679ae3)

**Comment by seifeldinio** [comment]
I've explored this solution for quick-editing within the inspector panel, here's the prototype for it:

https://github.com/user-attachments/assets/fc1b52a3-35e7-44db-931d-5d8fc8601754


Mockups:
### Single Page Selection

![Juste Une Page Selectionne](https://github.com/user-attachments/assets/6ec719f5-756f-423b-a6bd-30cf85a87959)

**Focus on Edit Modal:** When a user clicks on an option to edit, the table view lessens in opacity to focus on the edit modal. This guides the user's attention with a dark overlay, allowing them to focus on the content at hand.

![Une Page Selectionne - Status](https://github.com/user-attachments/assets/26d7c60f-de7f-426f-a91c-6500034c2925)

Editing options:

![Page Options](https://github.com/user-attachments/assets/4781685b-9c93-44ce-a9c8-b470f7c459fb)



### Multiple Page Selection

![Plusieur Pages Selectionnes](https://github.com/user-attachments/assets/a5d1e251-03ab-4d5b-9d69-cb443b7ad574)


![Plusieur Pages Selectionnes - Status](https://github.com/user-attachments/assets/3abca318-2196-4b4e-a15e-e5993585136a)


Editing options for multiple page selection:

![Multiple Pages Options](https://github.com/user-attachments/assets/889a9f56-45ce-440f-b5c8-6d8462058e8f)



### Quick Editing for Media Files

A similar quick-editing approach can be applied to media files

![Media selectionne](https://github.com/user-attachments/assets/3978ce8b-b69c-4fea-8b57-e6fdb4204910)

![Media selectionne - Alt Text](https://github.com/user-attachments/assets/6cd3ea29-62e5-4410-9307-198f717c0451)


Would love to hear your feedback on this!

**Comment by jasmussen** [comment]
This is looking pretty solid. A couple of details to figure out across the inspector itself, but high level that seems to capture the instincts I've had for this interaction. Would be good to hear from others in @WordPress/gutenberg-design as well!

**Comment by jasmussen** [comment]
Took a quick stab at revisiting this. Here's screenshots of what's experimentally shipping:

![Image](https://github.com/user-attachments/assets/6a64fd37-7e5d-4d12-ad34-659dfd8eb58f)

![Image](https://github.com/user-attachments/assets/7299847f-5a6d-49ac-b783-655656f77a3c)

_Noting there, by the way, that there's a horizontal scrollbar, which there probably shouldn't be._

The pieces are increasingly present, but some polish is missing. Here's a quick mockup:

![Image](https://github.com/user-attachments/assets/0fb8a99a-b540-4623-a9f9-b29d9d255b8e)

Mockup shows the quick edit panel very much resembling the page inspector, as well as a version with multiple items selected. Main changes to what's shipping in trunk:

* Changed the QuickEdit width to be the same as that of the page/template inspector. The value in having this be that much wider is unclear, and matching the width also helps the understanding: this is conceptually (and partially in practice), the same panel.
* Showed the regular Featured Image control. Since you'd only see this when you have a single item selected, it's unclear that this needs to be different.
* Changing a few metrics, notably the inner padding, and the ensuring the main title aligns vertically with the data view title to the left of it. 

What do you think? 

CC: @youknowriad @jorgefilipecosta @WordPress/gutenberg-design 

[Figma](https://www.figma.com/design/kEyWnuYWlskMGecBJcm92X/Quick-Edit?node-id=2-5027&t=A5YOLTSXv6GV9Eps-1).

**Comment by jasmussen** [comment]
Note, the above mockups are intentionally based on the same work and instincts [as these](https://github.com/WordPress/gutenberg/issues/55101#issuecomment-2267607005).

**Comment by youknowriad** [comment]
Looks good, I think some things still need to be thought about properly in terms of how to represent them 

- the "featured image"  field for instance: Current in trunk, it looks like that because it's a regular "media" field. Imagine you being able to add multiple kind of image fields to a given post (or entity). Would you want to render the image directly like that? Maybe it's ok :) we need to try it but we need to also think more generically. 

- It's unclear why "Last edited" is shown there, this is not really an "editable" field is it? so for a `DataForm` component, it's unclear why we show this information there, Maybe we need to support "readonly" fields that can be rendered within forms. (Again thinking what this represents in a more generic way)

**Comment by jasmussen** [comment]
> the "featured image" field for instance: Current in trunk, it looks like that because it's a regular "media" field. Imagine you being able to add multiple kind of image fields to a given post (or entity). Would you want to render the image directly like that? Maybe it's ok :) we need to try it but we need to also think more generically.

This was specifically a reflection of this field showing up _only_ when you select a single item. That field currently disappears when you select multiple. Thanks for the clarification, though. How's this?

![Image](https://github.com/user-attachments/assets/432f7db1-3d95-41a3-a229-0fe3066d8a2b)

That also addresses your other feedback about last edited. Thank you!

**Comment by youknowriad** [comment]
> This was specifically a reflection of this field showing up only when you select a single item. That field currently disappears when you select multiple. Thanks for the clarification, though. How's this?

It looks good, a couple of thoughts

 - Why the label is on top of the field for "media" but on the side for the "rest"? We can also remove the field if needed. For clarity, all of these options are currently supported (top, side, no visible label).
 - We don't have support for "mixed" values yet for bulk editing. I believe @louwie17 is exploring this but your design brings something new: the possibility to have a different "mixed" value representation depending on the field type. We might have to start without it (showing "Mixed" instead of the three images, but it's something to think about)

**Comment by jasmussen** [comment]
> Why the label is on top of the field for "media" but on the side for the "rest"? We can also remove the field if needed. For clarity, all of these options are currently supported (top, side, no visible label).

Practically, it's about scannability: the Media field is a general design tool like it is for Cover, Group, Site Logo and others. The other toggles are more like key/value pairs.

**Comment by youknowriad** [comment]
In the latest mockup it feels a bit weird design wise, in the sense that "Media" can also feel on the same level as "Projects" like a title for the whole thing.

**Comment by jasmussen** [comment]
Good point, can work without that label. It's a single panel, after all.

**Comment by jameskoster** [comment]
I'm also a little unsure about this media treatment, how would you know the association between file and field if they're grouped, and how would you edit the individual fields?

My thinking was that each would be a separate field, with `type` = `media`.

![Image](https://github.com/user-attachments/assets/de706905-233e-44f1-aec7-3551aec2d1ed)

The associated control could be an instance of [`FormFileUpload`](https://wordpress.github.io/gutenberg/?path=/docs/components-formfileupload--docs), likely integrated with the media library. That part needs some design work, but there are some rough mockups in https://github.com/WordPress/gutenberg/issues/65333.

**Comment by jasmussen** [comment]
Mighit work, though: "Downloadable file" wrapping doesn't work for me, and if we do move it to the right, I'd remove the border around it just like there's no border around "Status", which also has an icon.

**Comment by jameskoster** [comment]
> though: "Downloadable file" wrapping doesn't work for me

I don't like it either, but ultimately we don't control field names, they can be anything. This is a real example from WooCommerce. Translations are a consideration too. Perhaps we could truncate to one line, and show the full name via tooltip?

An alternative approach could be to commit to a more standard form layout / appearance, which would simplify the DataForms work quite a bit I imagine. 

> I'd remove the border around it just like there's no border around "Status", which also has an icon

The difference between these controls and the status button is that they're drop zones. I think it's beneficial to visually indicate that functionality somehow.

**Comment by jasmussen** [comment]
> I don't like it either, but ultimately we don't control field names, they can be anything. This is a real example from WooCommerce. Translations are a consideration too. Perhaps we could truncate to one line, and show the full name via tooltip?

Oh right, thank you for clarifying. We should not elide, and there's likely an action item for folks working on the woo product to optimize labels here so they capture that particular flavor of feature (e.g. "File download" or "Download" or "File" could be a suggestion for them, if they're reading!). 

> The difference between these controls and the status button is that they're drop zones. I think it's beneficial to visually indicate that functionality somehow.

🤔 Indeed. ItemGroup supports this, and a preview thumbnail on the left. It becomes a bit of a mix in this otherwise "key-value pair" layout, though. Should that particular type of control be surfaced in that same grid? 

Essentially this is a metadata panel, just like how Apple Photos or Adobe Lightroom have metadata panels, even while also having other very different layouts. I'll do a little digging and look at best practices across these; even each key/value pair is _metadata_, they are definitely different type of metadata, each of which might need its own affordances.

![Image](https://github.com/user-attachments/assets/17f6dc39-997d-4bbe-8e19-c670d8db2cb4)

**Comment by jameskoster** [comment]
> We should not elide, and there's likely an action item for folks working on the woo product to optimize labels here so they capture that particular flavor of feature (e.g. "File download" or "Download" or "File" could be a suggestion for them, if they're reading!).

I agree we should encourage plugin authors to use shorter field names where possible, but forcing them work around one specific UI feels like an overreach to me. Especially as that UI might change in the future. There will undoubtedly be situations where long field names cannot be avoided, or translations cause them indirectly. I think it would be better to find a design that adapts elegantly, if we can.

> It becomes a bit of a mix in this otherwise "key-value pair" layout, though. Should that particular type of control be surfaced in that same grid?

I see what you mean. I don't think the established convention is a great fit here though; the button would show the file name, and on click it would open a popover containing the UI to replace, remove, etc. That feels unnecessarily convoluted to me.

Worth noting we already have a mixture of behaviors in this 'panel' layout. In [this example](https://wordpress.github.io/gutenberg/?path=/docs/dataviews-dataform--docs) you can see that `booleans` appear as a toggle. Maybe it's okay for `media` fields to skip the popover?

To be honest I don't mind what is shipping, it just needs a label:

https://github.com/user-attachments/assets/1de3db73-d54d-4e77-b019-f55f951496c7

**Comment by jasmussen** [comment]
> There will undoubtedly be situations where long field names cannot be avoided, or translations cause them indirectly. I think it would be better to find a design that adapts elegantly, if we can.

I'm not sure what you're proposing here, is the eliding adapting elegantly, or do you mean a different UI altogether? In the case of this metadata I don't think it's unreasonable to list it like this, though as noted, I think there's wiggleroom in how different types of metadata gets listed, where some can be key/value pair style, potentially there's room for other tools. In fact I could argue: _that is what's shipping_, it's a different control for what _isn't_ a key/value pair. I'm here for that. But also to clarify, I don't think the wrapping is fundamentally problematic, if it wraps, it wraps, and this is handled. It was more a matter of: if that particular string came from core, we have an opportunity to do better. Small detail.

**Comment by jameskoster** [comment]
> I don't think the wrapping is fundamentally problematic, if it wraps, it wraps, and this is handled

Agreed :)

**Comment by youknowriad** [comment]
Can you do a quick TLDR of the discussion and add a quick todo list item to the issue description to clarify what's missing?

**Comment by jameskoster** [comment]
We were discussing how to treat media fields. Long term I feel they should probably render a comprehensive upload/dropzone component, but given we don't have that yet I think we can run with what we have. We just need to add a label. Whether that appears above or to the left of the control doesn't make too much difference imo.

<img width="376" alt="Image" src="https://github.com/user-attachments/assets/d5245e83-3427-4e2e-b14c-15ffcb32395f" />
<img width="374" alt="Image" src="https://github.com/user-attachments/assets/980a6caa-f563-4cf1-9ece-f1c3877906e7" />

Though if we go with the row orientation I'd adjust the label to "Choose image" to fix the eliding. 

---

Then there was your question about whether to display "Last edited" given it isn't a field, which we didn't discuss :) A couple of ideas;

1. Could `DataForms` support a `Summary` property which would be output above the form?
2. Change the revisions field, so that instead of showing 'Revisions: $count' it shows 'Last edited: $date', and clicking $date takes you to the revisions screen.

**Comment by youknowriad** [comment]
> Could DataForms support a Summary property which would be output above the form?

It could but if it's above the form, why should it be inside the component, it can just be outside but in this case, it's not above the form, it's within a form in a special position. IT looks more like a "readonly field" than a summary to me.

> Change the revisions field, so that instead of showing 'Revisions: $count' it shows 'Last edited: $date', and clicking $date takes you to the revisions screen.

What does "editing a revision" mean. For me even that one is more of "readonly field" regardless of how it's rendered. 

---

My question stands though, It would be cool to write down a todo list of the "mandatory things" before we can move this out of experimental.

**Comment by jameskoster** [comment]
> It could but if it's above the form, why should it be inside the component

The main reason to make it part of the component is to ensure consistency, and easy updates in the future (maybe we want to change the spacing or whatever). But I agree it's fine to say it's outside the component for now.

> For me even that one is more of "readonly field" regardless of how it's rendered.

You're saying both Revisions and Last modified should live outside of the form? That makes sense to me, but suppose it will require a small update to the existing document inspector since the revisions link currently lives in the 'form' section:

<img width="280" alt="Image" src="https://github.com/user-attachments/assets/6bac1bd7-694a-45d5-95ed-47455a40b9e0" />


> My question stands though, It would be cool to write down a todo list of the "mandatory things" before we can move this out of experimental.

Sorry I didn't realise you were asking generally :) In terms of functionality the main missing piece I see is bulk editing, but I know there are several PR's in the works to address that.

There are some design details we need to work out, but I don't know that they're blockers. The biggest one is better connecting the dialog to the form. Something about this pattern doesn't work as well outside the editor as it does inside:

<img width="1476" alt="Image" src="https://github.com/user-attachments/assets/c09ba780-df53-4ded-8f57-afb15851c5b4" />

Some spacing adjustments (panel width/padding) might help, but I'm curious about _trying_ the more traditional form layout.

**Comment by jasmussen** [comment]
> Some spacing adjustments (panel width/padding) might help, but I'm curious about trying the more traditional form layout.

It's a thing we can try, though so long as the appearance of the QuickEdit panel is the same (fundamentally) as that of the page inspector in the full editor, improvements we make affect both of them. At least in a systems sense, so it might make sense to take those two, together, as far as we can before starting to diverge. That's mainly an argument for width/padding changes first. Also perhaps scrimming the main canvas when opening a flyout. I think @auareyou also made a suggestion for the whole row itself to be a clickable button for the flyout, and having some sort of highlighted state when it opens being a benefit for both cases too.

Finally, there are some subtle pieces that seem missing between these two:

![Image](https://github.com/user-attachments/assets/c6f7d19b-c4a3-4cd1-8383-575612055135)

* It should just say "Status", not "Status & Visibility" (and even if it did say the latter, it should be "Status & visibility" as we use sentence case)
* The order is a bit off, with Author higher in this list than in the page inspector.
* Date works better when top-aligned, and the date itself wraps more nicely when the width of the overall panel is 280px. 

The main argument I've heard for being more close to a form, is that it saves a few clicks. But in most of these cases, it really doesn't, as these are dropdowns or popovers. It's really only "Slug" that would save a click, if it was an input right here.

**Comment by jameskoster** [comment]
> It's really only "Slug" that would save a click, if it was an input right here

Also Author, Date, Template, and Discussion. It's also easier for a form layout to adapt to different sizes which will be important if this panel ever becomes resizable in the future. The current format works less well at different widths, feeling too spaced out when wider, or running into wrapping issues when narrower.

<img width="1511" alt="Image" src="https://github.com/user-attachments/assets/182572a5-0787-4294-bbf0-a79c97e3e08a" />

Still, I agree we should explore how the existing design can be refined for this context. Making the whole row a single button has been suggested in the past as an a11y improvement so that sounds good to me. Here's a mockup, including the scrim idea and updated width/padding to match the Inspector:

https://github.com/user-attachments/assets/ff074339-4acb-4534-8ce4-1d09420811c6

If we're not confident about the dialog, modals could be another option:

https://github.com/user-attachments/assets/27a9cd12-20c4-4cce-ae1d-c0e63f6c59bc

Or even drilldowns, though I suspect this would be trickier to implement as part of the component:

https://github.com/user-attachments/assets/af9dce55-923e-4b0a-9bb7-0fc764a47371

**Comment by karmatosed** [comment]
@jameskoster and @jasmussen, there is a lot here and as this has the design feedback label, I wanted to be sure where and what that feedback was needed before side tracking.

To start, have any of these screens and thinking changed since December? If so, can we get an update, please, on this to ensure the feedback is aligned?

> In List layout it feels important that the preview remains visible.

I think having access to this makes sense, the point of ‘quick edit’ should be ease of access by default so anything that makes this visible is key.

I do have some feedback on a few pieces but before I went too far I wanted to be sure of not distracting. I am also tracking back on issues like https://github.com/WordPress/gutenberg/issues/70913

**Comment by jameskoster** [comment]
An initial version of this has been implemented as an experiment. At this point it's probably better to create dedicated issues for any outstanding pieces.
